network_simulator/flow.py

```diff
@@ -125,38 +173,428 @@ def handle_packet_loss(self, packet_id, loss_type, main_event_loop):
         :param MainEventLoop main_event_loop: main Event loop for further
         Event scheduling.
         """
-        pass
+        logger.debug("Flow %s lost packet %d due via PacketLossType %d",
+                     self.flow_id, packet_id, loss_type)
 
     @abstractmethod
-    def handle_packet_success(self, packet_id):
+    def handle_packet_success(self, packet, statistics, curr_time):
         """
         A handler for dealing with packets that successfully completed a
         round trip. This will adjust window sizes and make other state
-        changes, depending on the TCP congestion algorithm.
-        :param packet_id: ID of packet that completed a round trip
+        changes, depending on the TCP congestion algorithm. It will also
+        update common metadata and statistics.
+
+        The sending of additional Packets (e.g. if the window size is grown)
+        must be handled elsewhere, since this can be subclass-dependent.
+        :param Packet packet: The AckPacket that completed a round trip.
+        :param Statistics statistics: The Statistics to update.
+        :param float curr_time: The current simulation time (in seconds).
         """
-        pass
+        assert isinstance(packet, AckPacket)
```

Comment:
That is intended. Flows should only handle AckPackets; DataPackets are handled at-Host and the Host itself generates an ACK and sends it back. So if this condition is not met, the simulation should fail.

Even if it does fail, though, the AssertionError will trigger Statistics to be outputted (after we update MainEventLoop).